### PR TITLE
✨ [bento][amp-accordion] Add Imperative API for amp-accordion with Tests and Storybook

### PR DESCRIPTION
**Pull request by krdwan**
API Update 10/27

`toggle()` and `expand()` API methods called with no arguments for an accordion that has the attribute `expand-single-section` are special cases and are handled uniquely.  These two actions will not change the behavior of the accordion at all by definition.  These functions do not have a clearly defined behavior for accordions of more than 1 section.  To stay consistent, they will also not have an impact on accordions with 1 section.

****
Initial Description Below (10/20)

There are three API functions for Accordion in Preact mode

1. `toggle(id)` - if called with an `id`, the accordion will check for an Accordion section that has the same `id` and toggle it.  If called without an `id` it will toggle all sections.  Behavior is undefined for toggle all sections when `expand-single-section` is true.
2. `expand(id)` - if called with an `id`, the accordion will check for an Accordion section that has the same `id` and expand it.  If called without an `id` it will expand all sections.  Behavior is undefined for expand all sections when `expand-single-section` is true. 
3. `collapse(id)` - if called with an `id`, the accordion will check for an Accordion section that has the same `id` and collapse it.  If called without an `id` it will collapse all sections.  

Implementation Note:
This PR introduces a new state variable `realIdMap` which is a JSON object set by `useRef`.  This is a mapping from an accordion section's `id` prop to an internal `id` that is used by the accordion to track each section.  This is needed because the `toggle`, `expand`, `collapse` functions should only operate on a publisher provided `id` and not on internal `id`.  If multiple accordion sections have the same `id`, only the first one will be saved in the `realIdMap`.

Outstanding Issues:
1. `expandedMap` must be included in the dependency list of `useImperativeHandle` or else functionality is not as expected.  It appears that the functions do not get updated with the `expandedMap` unless `expandedMap` is included in the dependency list.  The compiler does not seem to like that `expandedMap` is in the dependency list (jump to line https://github.com/ampproject/amphtml/pull/30754/files#diff-29014805b49be0bc534bc1ea1a75299e796116cc8c26d01fa1e2042ac2107f26R94)
2. behavior still not fully working when `expand-single-section` is true

### CODE REVIEW

File: extensions/amp-accordion/1.0/accordion.js [hunk 1]
```diff
@@ -82,6 +88,18 @@ export function Accordion({
     });
   }, [expandSingleSection]);
 
+  useImperativeHandle(
+    ref,
+    () => {
+      return {
+        registerSection,
```
Comment (dvoytenko) [inline extensions/amp-accordion/1.0/accordion.js:95]: Just one note that, `toggleExpanded` and `isExpanded` could be right here. But `registerSection` is definitely out of place. Think about it this way: the imperative API (along with component props) is part of the component's public API. The `registerSection` is an internal detail.

Comment (krdwan) [reply extensions/amp-accordion/1.0/accordion.js]: Definitely makes sense, thanks Dima!

File: extensions/amp-accordion/1.0/accordion.js [hunk 2]
```diff
@@ -177,9 +261,9 @@ export function AccordionSection({
 
   useLayoutEffect(() => {
     if (registerSection) {
-      return registerSection(id, defaultExpanded);
+      return registerSection(id, defaultExpanded, propId);
```
Comment (dvoytenko) [inline extensions/amp-accordion/1.0/accordion.js:264]: Can we just use real ID instead of auto-gen ID when specified?

File: extensions/amp-accordion/1.0/accordion.js [hunk 3]
```diff
@@ -156,6 +239,7 @@ export function AccordionSection({
   children,
   ...rest
 }) {
+  const propId = rest['id'];
```
Comment (dvoytenko) [inline extensions/amp-accordion/1.0/accordion.js:242]: Should be just a normal documented prop. Can always be aliased when destructured or we can rename the old one. E.g.

```
Section({id: propId, ...}) {
  const [genId] = useState(generateSectionId);
  const id = propId || genId;
  ... the rest is the same ...
}
```

File: extensions/amp-accordion/1.0/accordion.js [hunk 4]
```diff
@@ -39,22 +41,29 @@ const AccordionContext = Preact.createContext(
...
+    expandSingleSection = false,
+    animate = false,
+    children,
+    id,
+    ...rest
+  },
+  ref
+) {
   const [expandedMap, setExpandedMap] = useState(EMPTY_EXPANDED_MAP);
+  const realIdMap = useRef(EMPTY_REAL_ID_MAP);
```
Comment (dvoytenko) [inline extensions/amp-accordion/1.0/accordion.js:66]: Per note below, I think this can be avoided entirely.

File: extensions/amp-accordion/1.0/accordion.js [hunk 5]
```diff
@@ -72,8 +81,73 @@ export function Accordion({
     });
   }, [expandSingleSection]);
 
+  useImperativeHandle(
+    ref,
+    () => {
+      return {
```
Comment (dvoytenko) [inline extensions/amp-accordion/1.0/accordion.js:87]: inline

File: extensions/amp-accordion/1.0/accordion.js [hunk 6]
```diff
@@ -72,8 +81,73 @@ export function Accordion({
     });
   }, [expandSingleSection]);
 
+  useImperativeHandle(
```
Comment (dvoytenko) [inline extensions/amp-accordion/1.0/accordion.js:84]: IMHO, the `const toggle = ` and other declarations should precede this call.

File: extensions/amp-accordion/1.0/accordion.js [hunk 7]
```diff
@@ -72,8 +81,73 @@ export function Accordion({
...
+    [toggle, collapse, expand, expandedMap]
+  );
+
+  const toggle = useCallback(
+    (id) => {
+      if (id) {
+        if (realIdMap.current[id]) {
+          toggleExpanded(realIdMap.current[id]);
+        }
+      } else {
```
Comment (dvoytenko) [inline extensions/amp-accordion/1.0/accordion.js:103]: Is this `else` really the part of our API? IMHO looks really unexpected. Same for all other `}else{`

File: extensions/amp-accordion/1.0/accordion.js [hunk 8]
```diff
@@ -72,8 +81,73 @@ export function Accordion({
     });
   }, [expandSingleSection]);
 
+  useImperativeHandle(
+    ref,
+    () => {
```
Comment (dvoytenko) [inline extensions/amp-accordion/1.0/accordion.js:86]: The `useImperativeHandle` result should be defined in our `.type.js`. Just a reminder. I realize this is an early draft.

File: extensions/amp-accordion/1.0/accordion.js (hunk 2 above)
Comment (krdwan) [reply extensions/amp-accordion/1.0/accordion.js]: Updated in latest revision.  Per offline conversation we will use real ID or auto-gen ID here.  This potentially allows user to access sections without explicit ID defined, but we will assume user will use component correctly or get undefined behavior.

File: extensions/amp-accordion/1.0/accordion.js (hunk 3 above)
Comment (krdwan) [reply extensions/amp-accordion/1.0/accordion.js]: Thanks, updated!

File: extensions/amp-accordion/1.0/accordion.js (hunk 4 above)
Comment (krdwan) [reply extensions/amp-accordion/1.0/accordion.js]: Removed.  Per off-line conversation allows API to directly access any section.  We will trust the users not to use the API incorrectly.

File: extensions/amp-accordion/1.0/accordion.js (hunk 6 above)
Comment (krdwan) [reply extensions/amp-accordion/1.0/accordion.js]: Thanks, updated!

File: extensions/amp-accordion/1.0/accordion.js (hunk 7 above)
Comment (krdwan) [reply extensions/amp-accordion/1.0/accordion.js:118]: Correct, this is part of the `0.1` API as discussed offline.

File: extensions/amp-accordion/1.0/accordion.js [hunk 9]
```diff
@@ -97,16 +104,82 @@ export function Accordion({
...
 
+  const isExpanded = useCallback(
+    (id, defaultExpanded) => expandedMap[id] ?? defaultExpanded,
+    [expandedMap]
+  );
+
+  const toggle = useCallback(
+    (id) => {
+      if (id) {
+        if (hasOwn(expandedMap, id)) {
```
Comment (dvoytenko) [inline extensions/amp-accordion/1.0/accordion.js:115]: They can only be possibly "own" properties here. So either `id in expandedMap` or `expandedMap[id] != null` are totally functional in this case.

File: extensions/amp-accordion/1.0/accordion.js [hunk 10]
```diff
@@ -97,16 +104,82 @@ export function Accordion({
...
+        }
+      }
+    },
+    [expandedMap, toggleExpanded, isExpanded]
+  );
+
+  useImperativeHandle(
+    ref,
+    () => {
+      return {
```
Comment (dvoytenko) [inline extensions/amp-accordion/1.0/accordion.js:164]: This `return` can be inlined.

File: extensions/amp-accordion/1.0/accordion.js [hunk 11]
```diff
@@ -152,11 +229,12 @@ export function AccordionSection({
   animate: defaultAnimate = false,
   headerClassName = '',
   contentClassName = '',
+  id,
   header,
   children,
   ...rest
 }) {
-  const [id] = useState(generateSectionId);
+  const [internalId] = useState(id || generateSectionId);
```
Comment (dvoytenko) [inline extensions/amp-accordion/1.0/accordion.js:237]: This is clever, but it's better to do

```
const [genId] = useState(generateSectionId);
const internalId = id || genId;
```

This is better b/c the change of `id` is immediately reflected.

**Review by dvoytenko — CHANGES_REQUESTED** [review]

**Comment by dvoytenko** [comment]
@krdwan I made a few non-essential comments. This is ready to go out of draft and support tests and AMP bindings.

File: extensions/amp-accordion/1.0/accordion.js (hunk 8 above)
Comment (krdwan) [reply extensions/amp-accordion/1.0/accordion.js]: Thanks Dima, updated in latest revision!

File: extensions/amp-accordion/1.0/accordion.js (hunk 5 above)
Comment (krdwan) [reply extensions/amp-accordion/1.0/accordion.js]: Thanks, Dima, updated in latest revision!

File: extensions/amp-accordion/1.0/accordion.js (hunk 10 above)
Comment (krdwan) [reply extensions/amp-accordion/1.0/accordion.js]: Thanks, updated in latest!

File: extensions/amp-accordion/1.0/accordion.js (hunk 9 above)
Comment (krdwan) [reply extensions/amp-accordion/1.0/accordion.js]: Thank you, very helpful, updated in the latest!

File: extensions/amp-accordion/1.0/accordion.js (hunk 11 above)
Comment (krdwan) [reply extensions/amp-accordion/1.0/accordion.js]: Thanks for the suggestion, updated!

**Comment by krdwan** [comment]
> @krdwan I made a few non-essential comments. This is ready to go out of draft and support tests and AMP bindings.

Thanks @dvoytenko out of draft now including AMP bindings and tests!

File: extensions/amp-accordion/1.0/amp-accordion.js [hunk 12]
```diff
@@ -34,9 +35,29 @@ const HEADER_SHIM_PROP = '__AMP_H_SHIM';
...
 
+/** @extends {PreactBaseElement<AccordionDef.AccordionApi>} */
 class AmpAccordion extends PreactBaseElement {
   /** @override */
   init() {
+    this.registerApiAction(
+      'toggle',
+      (api, invocation) =>
+        api.toggle(invocation.args && invocation.args['section']),
+      ActionTrust.LOW
```
Comment (dvoytenko) [inline extensions/amp-accordion/1.0/amp-accordion.js:46]: Are you sure it's a `LOW` trust here? I'd expect at least a default trust level.

File: extensions/amp-accordion/1.0/test/test-accordion.js [hunk 13]
```diff
@@ -459,4 +457,198 @@ describes.sandboxed('Accordion preact component', {}, (env) => {
...
+    let wrapper;
+    let ref;
+
+    let section1;
+    let section2;
+    let section3;
+
+    describe('multi-expand accordion', () => {
+      beforeEach(() => {
+        ref = Preact.useRef();
```
Comment (dvoytenko) [inline extensions/amp-accordion/1.0/test/test-accordion.js:471]: I think in this case, it should be `Preact.createRef()`. `useRef` is only for use inside the components themselves.

**Review by dvoytenko — CHANGES_REQUESTED** [review]
Looks great. Just a couple of comments from me.

File: extensions/amp-accordion/1.0/amp-accordion.js (hunk 12 above)
Comment (krdwan) [reply extensions/amp-accordion/1.0/amp-accordion.js]: Thanks Dima!  Per offline conversation, removing `ActionTrust` to use the default value in `registerAction` which is `ActionTrust.DEFAULT`.

File: extensions/amp-accordion/1.0/test/test-accordion.js (hunk 13 above)
Comment (krdwan) [reply extensions/amp-accordion/1.0/test/test-accordion.js]: Thanks, updated!

File: extensions/amp-accordion/1.0/amp-accordion.js (hunk 12 above)
Comment (dvoytenko) [reply extensions/amp-accordion/1.0/amp-accordion.js]: That's good. But I think you need to modify `registerApiAction` declaration in the `PreactBaseElement` slightly. Otherwise it will complain about a missing arg. I think you can just add `=` to the `minTrust` type. Run `gulp check-types` to check.

Comment (krdwan) [reply extensions/amp-accordion/1.0/amp-accordion.js]: Thanks Dima, this is updated.  I also refactored another issue I saw when running `gulp check-types`

`id in expandedMap && !isExpanded(id)` updated to => `!isExpanded(id, true)`
`id in expandedMap && isExpanded(id)` updated to => `isExpanded(id, false)`

this is because isExpanded actually requires two arguments, and already protects for `id in expandedMap`.  Here it is for your reference:
isExpanded = `(id, defaultExpanded) => expandedMap[id] ?? defaultExpanded,`

Summary of the latest changes: https://github.com/ampproject/amphtml/pull/30754/commits/d16497d362547ce302d4ef917e82ed2fb2462ec0

**Comment by krdwan** [comment]
@dvoytenko @caroqliu Heads up, I've made a few updates to pass `gulp presubmit`.  `gulp presubmit` does not like a few forbidden strings in the code including `.collapse(` and `.expand(`.  I had to do a bit of a workout around in `Basic.amp.js`.

Here's the summary of changes: https://github.com/ampproject/amphtml/pull/30754/commits/1c77fccb3e036e698b0a8cbe7b074cf8fcd9ab91

Here's the forbidden strings item for reference: https://github.com/ampproject/amphtml/blob/master/build-system/tasks/presubmit-checks.js#L951

**Review by dvoytenko — APPROVED** [review]

File: extensions/amp-accordion/1.0/accordion.js [hunk 14]
```diff
@@ -97,16 +103,87 @@ export function Accordion({
...
+  );
+
+  const toggle = useCallback(
+    (id) => {
+      if (id) {
+        if (id in expandedMap) {
+          toggleExpanded(id);
+        }
+      } else {
+        // Toggle all should do nothing when expandSingleSection is true
```
Comment (caroqliu) [inline extensions/amp-accordion/1.0/accordion.js:118]: From manual testing on 0.1, it looks like when `expandSingleSection` is true, `toggle` without `id` will expand the last collapsed section. I think this is a side effect (iterate and toggle all elements, each new open one closing the last) rather than an actual intended API.

We should feel comfortable breaking the "toggle without given `id`" interaction pattern smartly IMO, though it's not urgent to break away from status quo if it's not needed. Can we start a thread with UX about if there are any best practices/expected interactions when a user toggles accordion sections, so we can make an informed decision in the long term? At the very least we can improve our documentation on what users should expect.

File: extensions/amp-accordion/1.0/accordion.js [hunk 15]
```diff
@@ -97,16 +103,87 @@ export function Accordion({
     [expandSingleSection]
   );
 
+  const isExpanded = useCallback(
+    (id, defaultExpanded) => expandedMap[id] ?? defaultExpanded,
+    [expandedMap]
+  );
+
+  const toggle = useCallback(
```
Comment (caroqliu) [inline extensions/amp-accordion/1.0/accordion.js:111]: Suggestion: `toggle`, `collapse`, and `expand` all look nearly identical in their implementations, is it possible to dedup some of this logic?

File: extensions/amp-accordion/1.0/amp-accordion.js [hunk 16]
```diff
@@ -34,9 +34,20 @@ const HEADER_SHIM_PROP = '__AMP_H_SHIM';
...
 
+/** @extends {PreactBaseElement<AccordionDef.AccordionApi>} */
 class AmpAccordion extends PreactBaseElement {
   /** @override */
   init() {
+    this.registerApiAction('toggle', (api, invocation) =>
+      api./*OK*/ toggle(invocation.args && invocation.args['section'])
+    );
+    this.registerApiAction('expand', (api, invocation) =>
+      api./*OK*/ expand(invocation.args && invocation.args['section'])
```
Comment (caroqliu) [inline extensions/amp-accordion/1.0/amp-accordion.js:45]: Will you be adding firing AMP events in a separate PR?

**Review by caroqliu — APPROVED** [review]

File: extensions/amp-accordion/1.0/amp-accordion.js (hunk 16 above)
Comment (dvoytenko) [reply extensions/amp-accordion/1.0/amp-accordion.js:45]: Thanks @caroqliu! It's an important topic. If a separate PR, let's make sure we track.

File: extensions/amp-accordion/1.0/accordion.js (hunk 15 above)
Comment (krdwan) [reply extensions/amp-accordion/1.0/accordion.js:111]: Thanks for the suggestion.  I also felt this way initially, and considered writing some type of generic function in which you can pass in the specific behavior per each different action.  Ultimately, I decided this looks a lot cleaner in terms of readability.  Especially since we actually do have 6 distinct behaviors (2 for each of the 3 actions).

Also as a bonus, later if we have to re-factor the behavior of `expandSingleSection` I believe it's easier in the current format.

Thoughts?

File: extensions/amp-accordion/1.0/accordion.js (hunk 14 above)
Comment (krdwan) [reply extensions/amp-accordion/1.0/accordion.js:118]: Right, I couldn't definitively describe the expected behavior of `toggle` and `expand` when `expandSingleSection` is true while testing in `0.1`.  There's also the use case of 1 vs 2 vs 2+ sections in the accordion, each which may have distinct behavior.  I also saw that it was not defined in the amp.dev, so to me, it felt like undefined behavior.  I can start a thread though.

Can we consider this non-blocking and address any feedback from UX in a separate PR?

File: extensions/amp-accordion/1.0/amp-accordion.js (hunk 16 above)
Comment (krdwan) [reply extensions/amp-accordion/1.0/amp-accordion.js:45]: Thank yes, this will be in a separate PR!

File: extensions/amp-accordion/1.0/accordion.js (hunk 14 above)
Comment (krdwan) [reply extensions/amp-accordion/1.0/accordion.js:118]: Per guidance from @nainar filed issue: https://github.com/ampproject/amphtml/issues/30895 and continue forward.

File: extensions/amp-accordion/1.0/accordion.js (hunk 15 above)
Comment (krdwan) [reply extensions/amp-accordion/1.0/accordion.js:111]: Since I haven't heard back, I'm going to go ahead and merge this PR.  But if you have additional thoughts we can address in a follow up.  Thanks!

Comment (caroqliu) [reply extensions/amp-accordion/1.0/accordion.js:111]: SGTM

Comment (krdwan) [reply extensions/amp-accordion/1.0/accordion.js:111]: Great!
